qrexec/policy/parser.py

```diff
@@ -0,0 +1,1543 @@
+            and (self.argument is None or self.argument == request.argument)
+            and self.source.match(request.source,
+                    system_info=request.system_info))
+
+
+class AbstractParser(metaclass=abc.ABCMeta):
+    '''A minimal, pluggable, validating policy parser'''
+
+    #: default rule type
+    rule_type = Rule
+
+    @staticmethod
+    def _fix_filepath(file, filepath):
+        if filepath is None:
+            try:
+                filepath = pathlib.Path(file.name)
+            except AttributeError:
+                if isinstance(file, io.IOBase):
+                    filepath = '<buffer>'
+        return file, filepath
+
+    def load_policy_file(self, file, filepath):
+        '''Parse a policy file'''
+        file, filepath = self._fix_filepath(file, filepath)
+
+        for lineno, line in enumerate(file, start=1):
+            line = line.strip()
+
+            # skip empty lines and comments
+            if not line or line[0] == '#':
+                self.handle_comment(line, filepath=filepath, lineno=lineno)
+                continue
+
+            if line.startswith('!'):
+                directive, *params = line.split()
+
+                if directive == '!include':
+                    try:
+                        included_path, = params
+                    except ValueError:
+                        raise PolicySyntaxError(filepath, lineno,
+                            'invalid number of params')
+                    self.handle_include(pathlib.PurePosixPath(included_path),
+                        filepath=filepath, lineno=lineno)
+                    continue
+                if directive == '!include-dir':
+                    try:
+                        included_path, = params
+                    except ValueError:
+                        raise PolicySyntaxError(filepath, lineno,
+                            'invalid number of params')
+                    self.handle_include_dir(
+                        pathlib.PurePosixPath(included_path),
+                        filepath=filepath, lineno=lineno)
+                    continue
+                if directive == '!include-service':
+                    try:
+                        service, argument, included_path = params
+                    except ValueError:
+                        raise PolicySyntaxError(filepath, lineno,
+                            'invalid number of params')
+                    self.handle_include_service(service, argument,
+                        pathlib.PurePosixPath(included_path),
+                        filepath=filepath, lineno=lineno)
+                    continue
+
+                if directive == '!compat-4':
+                    if params:
+                        raise PolicySyntaxError(filepath, lineno,
+                            'invalid number of params')
+                    logging.warning(
+                        'warning: !compat directive in file %s line %s'
```

Comment:
```suggestion
                        'warning: !compat-4 directive in file %s line %s'
```

But then, I'm not really sure if `!compat-4` is a good idea, IMO `!compat` would be better. Because it is R4.1, so `!compat-4` may be confusing.

Reply:
What if we need another `!compat` in the future? I'd like to avoid situation in which we find 10-years old policy with `!compat` and we don't know, what is compatible with what, which is also confusing.

Would `!compat-4.0` be OK? I can see dangerous similarity to `!version`, which AFAIR was decided against, but this would solve clarity problem.

Reply:
`!compat-4.0` would be definitely better. Alternative would be some descriptive name, like `!compat-per-service-policy`, but I'm not insisting on that.

Reply:
No, I think `!compat-<version>` should be sufficient, for some precision of version. If anything, I think it would be OK to plan for only single compat statement in whole policy, which would include old policy in its entirety. Descriptive name suggests there may be some other compat statement. I aim for only one compat support at any given time.